File: queens.py
import time

def display(GRID, QUEENS):
    N = 8
    CELL_WIDTH = 4  # adjust as needed for spacing

    ROWS = [GRID[i * N:(i + 1) * N] for i in range(N)]
    QUEEN_POSITIONS = [QUEENS[i * N:(i + 1) * N] for i in range(N)]

    def get_corner(i, row, previous_row):
        if previous_row is None:
            return "+" if i == 0 or row[i] != row[i - 1] else "-"
        if previous_row[i - 1] == row[i - 1] and previous_row[i] == row[i] and row[i - 1] == row[i]:
            return " "
        if i == 0 or previous_row[i - 1] == row[i - 1] and previous_row[i] == row[i]:
            return "|"
        if previous_row[i] == previous_row[i - 1] and row[i] == row[i - 1]:
            return "-"
        return "+"

    def print_border(row, previous_row):
        border = ""
        for i in range(N):
            corner = get_corner(i, row, previous_row)
            if previous_row is None or previous_row[i] != row[i]:
                border += corner + "-" * CELL_WIDTH
            else:
                border += corner + " " * CELL_WIDTH
        border += "+" if previous_row is None or row[-1] != previous_row[-1] else "|"
        print(border)

    def print_content(row, queen_row):
        line = "|"
        for i, cell in enumerate(row):
            line += " Q " + " " if queen_row[i] == 'Q' else " " * CELL_WIDTH
            if i < N - 1:
                line += "|" if row[i] != row[i + 1] else " "
        line += "|"
        print(line)

    def print_grid(rows, queen_positions):
        previous_row = None
        for row, queen_row in zip(rows, queen_positions):
            print_border(row, previous_row)
            print_content(row, queen_row)
            previous_row = row
        print_final_border(rows[-1])

    def print_final_border(last_row):
        line = "+" + "-" * CELL_WIDTH
        for i in range(1, N):
            if last_row[i - 1] == last_row[i]:
                line += "-" * (CELL_WIDTH + 1)
            else:
                line += "+" + "-" * CELL_WIDTH
        print(line + "+")

    print_grid(ROWS, QUEEN_POSITIONS)

def find_empty_colour(GRID, QUEENS):
    colours = set(GRID)
    sorted_colours = sorted(colours, key=lambda c: GRID.count(c))
    for colour in sorted_colours:
        found = True
        for i in range(len(GRID)):
            if GRID[i] == colour and QUEENS[i] == 'Q':
                found = False
                break
        if found:
            return colour
    return None

def is_valid(GRID, QUEENS, i, colour):
    N = 8
    # Check cell matches colour
    if GRID[i] != colour:
        return False
    
    # Check no queens in the same colour
    for j in range(len(GRID)):
        if QUEENS[j] == 'Q' and GRID[j] == GRID[i]:
            return False

    # Check no queens in the same row
    if 'Q' in QUEENS[i - i % N:i - i % N + N]:
        return False

    # Check no queens in the same column
    if 'Q' in QUEENS[i % N::N]:
        return False

    # Check no two queens in a 3x3 area
    row, col = divmod(i, N)
    for r in range(row - 1, row + 2):
        for c in range(col - 1, col + 2):
            if 0 <= r < N and 0 <= c < N and QUEENS[r * N + c] == 'Q':
                return False

    return True

def solve(GRID, QUEENS):
    colour = find_empty_colour(GRID, QUEENS)

    if colour is None:
        return True
    
    for i in range(len(GRID)):
        if is_valid(GRID, QUEENS, i, colour):
            QUEENS[i] = 'Q'

            if solve(GRID, QUEENS):
                return True

            QUEENS[i] = '.'  # backtrack
    return False

GRID = list('1111222213334242113444441134545413335456177758561717555611177666')
QUEENS = list('................................................................')

# Main execution
display(GRID, QUEENS)
start_time = time.time()
if solve(GRID, QUEENS):
    print("Solution found in %.2f seconds" % (time.time() - start_time))
    display(GRID, QUEENS)
else:
    print("No solution found")

File: queens.js
function display(GRID, QUEENS) {
    const N = 8;
    const CELL_WIDTH = 3; // adjust as needed for spacing

    // ...existing code: create rows arrays...
    const rows = [];
    const queenPositions = [];
    for (let i = 0; i < N; i++) {
        rows.push(GRID.slice(i * N, (i + 1) * N));
        queenPositions.push(QUEENS.slice(i * N, (i + 1) * N));
    }

    function getCorner(i, row, previousRow) {
        if (previousRow === null) {
            if (i === 0) return "┌";
            return row[i] !== row[i - 1] ? "┬" : "─";
        }
        if (i === 0) return "│";

        let topLeft = previousRow[i - 1], topRight = previousRow[i];
        let bottomLeft = row[i - 1], bottomRight = row[i];

        if (topLeft === bottomLeft && topLeft === topRight && topRight === bottomRight) return " ";
        if (topLeft === topRight && topRight === bottomLeft) return "┌";
        if (topLeft === topRight && topRight === bottomRight) return "┐";
        if (bottomLeft === bottomRight && bottomLeft === topLeft) return "└";
        if (bottomLeft === bottomRight && bottomLeft === topRight) return "┘";
        if (topLeft === bottomLeft && topRight !== bottomRight) return "├";
        if (topRight === bottomRight && topLeft !== bottomLeft) return "┤";
        if (bottomLeft === bottomRight && topLeft !== topRight) return "┴";
        if (topLeft === topRight && bottomLeft !== bottomRight) return "┬";
        if (topLeft === bottomLeft && topRight === bottomRight) return "│";
        if (topLeft === topRight && bottomLeft === bottomRight) return "─";
        return "┼";
    }

    function printBorder(row, previousRow) {
        let border = "";
        for (let i = 0; i < N; i++) {
            let corner = getCorner(i, row, previousRow);
            if (previousRow === null || previousRow[i] !== row[i]) {
                border += corner + "─".repeat(CELL_WIDTH);
            } else {
                border += corner + " ".repeat(CELL_WIDTH);
            }
        }
        if (previousRow === null) {
            border += "┐";
        } else if (row[N - 1] !== previousRow[N - 1]) {
            border += "┤";
        } else {
            border += "│";
        }
        console.log(border);
    }

    function printContent(row, queenRow) {
        let line = "│";
        for (let i = 0; i < N; i++) {
            line += queenRow[i] === 'Q' ? "Q".padStart(Math.floor((CELL_WIDTH + 1) / 2)).padEnd(CELL_WIDTH) : " ".repeat(CELL_WIDTH);
            if (i < N - 1) line += row[i] !== row[i + 1] ? "│" : " ";
        }
        line += "│";
        console.log(line);
    }

    function printFinalBorder(lastRow) {
        let line = "└" + "─".repeat(CELL_WIDTH);
        for (let i = 1; i < N; i++) {
            if (lastRow[i - 1] === lastRow[i]) {
                line += "─".repeat(CELL_WIDTH + 1);
            } else {
                line += "┴" + "─".repeat(CELL_WIDTH);
            }
        }
        console.log(line + "┘");
    }

    function printGrid(rows, queenPositions) {
        let previousRow = null;
        for (let j = 0; j < rows.length; j++) {
            printBorder(rows[j], previousRow);
            printContent(rows[j], queenPositions[j]);
            previousRow = rows[j];
        }
        printFinalBorder(rows[rows.length - 1]);
    }

    printGrid(rows, queenPositions);
}

function findEmptyColour(GRID, QUEENS) {
    const colours = Array.from(new Set(GRID));
    colours.sort((a, b) => GRID.filter(c => c === a).length - GRID.filter(c => c === b).length);
    for (let colour of colours) {
        let found = true;
        for (let i = 0; i < GRID.length; i++) {
            if (GRID[i] === colour && QUEENS[i] === 'Q') {
                found = false;
                break;
            }
        }
        if (found) return colour;
    }
    return null;
}

function isValid(GRID, QUEENS, i, colour) {
    const N = 8;
    if (GRID[i] !== colour) return false;

    for (let j = 0; j < GRID.length; j++) {
        if (QUEENS[j] === 'Q' && GRID[j] === GRID[i]) return false;
    }

    // Check row
    const rowStart = i - (i % N);
    if (QUEENS.slice(rowStart, rowStart + N).includes('Q')) return false;

    // Check column
    for (let j = i % N; j < GRID.length; j += N) {
        if (QUEENS[j] === 'Q') return false;
    }

    // Check 3x3 neighbourhood
    const row = Math.floor(i / N), col = i % N;
    for (let r = row - 1; r <= row + 1; r++) {
        for (let c = col - 1; c <= col + 1; c++) {
            if (r >= 0 && r < N && c >= 0 && c < N && QUEENS[r * N + c] === 'Q') {
                return false;
            }
        }
    }
    return true;
}

function solve(GRID, QUEENS) {
    const colour = findEmptyColour(GRID, QUEENS);
    if (colour === null) return true;

    for (let i = 0; i < GRID.length; i++) {
        if (isValid(GRID, QUEENS, i, colour)) {
            QUEENS[i] = 'Q';
            if (solve(GRID, QUEENS)) return true;
            QUEENS[i] = '.'; // backtrack
        }
    }
    return false;
}

// ...existing code: initialize GRID and QUEENS...
let GRID = '1111222213334242113444441134545413335456177758561717555611177666'.split('');
let QUEENS = '................................................................'.split('');

// Main execution
display(GRID, QUEENS);
let startTime = Date.now();
if (solve(GRID, QUEENS)) {
    console.log("Solution found in " + ((Date.now() - startTime) / 1000).toFixed(2) + " seconds");
    display(GRID, QUEENS);
} else {
    console.log("No solution found");
}

export { solve, display };
